src/main.cpp

```diff
@@ -28,20 +27,47 @@ int Main(vector<string> args)
 		%whitespace <- [ \t]*
 	)";
 
+	// create parser
 	parser parser(syntax);
 
 	if (!parser){
 		cout << "error creating grammar" << endl;
 		return EXIT_FAILURE;
 	}
 
-	parser.enable_packrat_parsing(); // Enable packrat parsing.
+	// create callbacks for debugging
+	auto rules = parser.get_rule_names();
 
-	int val;
+	for(auto r:rules){
+		parser[r.c_str()].enter = [r](const char* s, size_t n, any& dt) {
+			auto& indent = *dt.get<int*>();
+			for(int i=0; i<indent; i++){cout << "|  ";}
+			string beginning(s, n);
+			if(beginning.length()>10){beginning = beginning.substr(0,10) + "...";}
+			cout << r << " => \"" << beginning << "\"?" << endl;
+			indent++;
+		};
+
+		parser[r.c_str()].leave = [r](const char* s, size_t n, bool match, any& dt) {
+			auto& indent = *dt.get<int*>();
+			indent--;
+			for(int i=0; i<indent; i++){cout << "|  ";}
+			if(match){cout << "`-> match" << endl;}
+			else{cout << "`-> failed" << endl;}
```

Comment:
nit: stylistic question, but this could be

    cout << "`-> " << (match ? "match" : "failed") << endl;